### PR TITLE
fix UBs detected by UBSan

### PR DESCRIPTION
**Pull request by gfx**
All of them are the same type of UB: "applying (non-)zero offset to pointer".

### CODE REVIEW

File: lib/core/logconf.c [hunk 1]
```diff
@@ -407,8 +407,10 @@ static inline char *append_safe_string(char *pos, const char *src, size_t len)
 
 static char *append_unsafe_string_apache(char *pos, const char *src, size_t len)
 {
-    const char *src_end = src + len;
+    if (len == 0)
+        return pos;
 
+    const char *src_end = src + len;
     for (; src != src_end; ++src) {
```
Comment (kazuho) [inline lib/core/logconf.c:414]:
```suggestion
    for (const char *src_end = src + len; src != src_end; ++src) {
```

File: lib/core/proxy.c [hunk 2]
```diff
@@ -198,9 +198,9 @@ static void build_request(h2o_req_t *req, h2o_iovec_t *method, h2o_url_t *url, h
 
     /* headers */
     h2o_iovec_vector_t cookie_values = {NULL};
-    {
+    int found_early_data = 0;
+    if (H2O_LIKELY(req->headers.size != 0)) {
         const h2o_header_t *h, *h_end;
-        int found_early_data = 0;
         for (h = req->headers.entries, h_end = h + req->headers.size; h != h_end; ++h) {
```
Comment (kazuho) [inline lib/core/proxy.c:204]: Can we do `for (const h2o_header_t *h = req->headers.entries, *h_end = h + req->headers.size; ...` here as well?

File: lib/core/request.c [hunk 3]
```diff
@@ -327,7 +327,7 @@ void h2o_dispose_request(h2o_req_t *req)
 
     h2o_timer_unlink(&req->_timeout_entry);
 
-    if (req->pathconf != NULL) {
+    if (req->pathconf != NULL && req->num_loggers != 0) {
         h2o_logger_t **logger = req->loggers, **end = logger + req->num_loggers;
         for (; logger != end; ++logger) {
```
Comment (kazuho) [inline lib/core/request.c:332]:
```suggestion
        for (h2o_logger_t **logger = req->loggers, **end = logger + req->num_loggers; logger != end; ++logger) {
```

File: lib/http3/server.c [hunk 4]
```diff
@@ -884,27 +884,29 @@ static void handle_buffered_input(struct st_h2o_http3_server_stream_t *stream, i
...
-            const char *err_desc = NULL;
-            if ((err = stream->recvbuf.handle_input(stream, &src, src_end, in_generator, &err_desc)) != 0) {
-                if (err == H2O_HTTP3_ERROR_INCOMPLETE) {
-                    if (!quicly_recvstate_transfer_complete(&stream->quic->recvstate))
-                        break;
-                    err = H2O_HTTP3_ERROR_GENERAL_PROTOCOL;
-                    err_desc = "incomplete frame";
+        if (bytes_available != 0) {
+            const uint8_t *src = (const uint8_t *)stream->recvbuf.buf->bytes, *src_end = src + bytes_available;
+            while (src != src_end) {
```
Comment (kazuho) [inline lib/http3/server.c:889]: Maybe we want to use do-while knowing that `bytes_available` is non-zero?

**Review by kazuho — COMMENTED** [review]
Thank you for the PR. Looks good, all my comments are editorial.

File: lib/common/url.c [hunk 5]
```diff
@@ -119,40 +119,42 @@ static h2o_iovec_t rebuild_path(h2o_mem_pool_t *pool, const char *src, size_t sr
...
 h2o_iovec_t h2o_url_normalize_path(h2o_mem_pool_t *pool, const char *path, size_t len, size_t *query_at, size_t **norm_indexes)
 {
-    const char *p = path, *end = path + len;
     h2o_iovec_t ret;
 
     *query_at = SIZE_MAX;
     *norm_indexes = NULL;
 
-    if (len == 0) {
+    if (len == 0 || (len == 1 && path[0] == '/')) {
```
Comment (kazuho) [inline lib/common/url.c:127]: Do we need this change?

I think the code below would run fast when the input is "/". Basically, it tests if `path[0]` is `/` then skip the for loop. That is exactly the same as the or clause added here.

File: lib/common/url.c [hunk 6]
```diff
@@ -119,40 +119,42 @@ static h2o_iovec_t rebuild_path(h2o_mem_pool_t *pool, const char *src, size_t sr
...
+                goto Return;
+            }
         }
-    }
 
-Return:
-    ret.base = (char *)path;
-    ret.len = p - path;
-    return ret;
+    Return:
```
Comment (kazuho) [inline lib/common/url.c:153]: Generally speaking, we only allow jumps to the end of the function. This change is an violation to that rule.

I get the idea of limiting the scope of variables, but IMO it is more important to not have code jumping to arbitrary locations.

Comment (gfx) [reply lib/common/url.c:153]: Agreed. Reverting.

File: lib/common/url.c (hunk 5 above)
Comment (gfx) [reply lib/common/url.c:127]: Oh, it makes sense. Thanks for the explanation. Reverting.

**Pull request merged**